doc/stklos.1.in
.\" -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH STKLOS 1 "version @VERSION@"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
stklos \- An efficient Scheme System providing a powerful Object
System

.SH SYNOPSYS
\fbstklos\fR [OPTION ...] [ARGUMENT ... ]

.SH DESCRIPTION
.B STklos
is a free Scheme System (nearly) conforming to R7RS. The aim of this
implementation is to be fast as well as light. The implementation is
based on an ad-hoc Virtual Machine.
.LP
The salient points of STklos are:
.IP "" 4
- an efficient and powerful object system based on CLOS providing:
.IP "" 8
+ Multiple Inheritance,
.IP "" 8
+ Generic Functions,
.IP "" 8
+ Multi-methods
.IP "" 8
+ an efficient MOP (Meta Object Protocol)
.IP "" 4
- a simple to use module system
.IP "" 4
- implementing the full tower of numbers defined in R7RS
.IP "" 4
- an easy connection to the GTK+ toolkit
.IP "" 4
- Perl compatible regular expressions thanks to the PCRE package
.IP "" 4
- properly implemented tail recursive calls

.SH OPTIONS
Here are the options available in the version @VERSION@ of
.B stklos.

.IP "-l file, --load=file"
load 'file' before going interactive
.IP "-f file, --file=file"
use 'file' as program
.IP "-I dir"
prepend 'dir' to the directory path. This option can be used several
times.
.IP "-A dir"
append 'dir' to the directory path. This option can be used several
times.
.IP "--includes-use-load-path"
\fIInclude\fR and \fIinclude-ci\fR use the \fIload-path\fR parameter object.
.IP "-e sexpr, --execute=sexpr"
evaluate the given sexpr and exit
.IP "-b file, --boot-file=file"
use 'file' to boot the system instead of the default one
.IP "-D directory, --conf-dir=directory"
use the given directory to store the various configuration files. This
directory is also used to store the packages installed with the
\fIstklos-pkg(1)\fR command. Note that the shell variable
\fISTKLOS_CONFDIR\fR can also be used to set this directory (see below).
.IP "-q, --no-init-file"
quiet: do not load the user init file.
.IP "-Q, --no-startup-message"
quiet: do not show the startup message.
.IP "-i, --interactive"
be interactive, even if the command finds it is not the case. This
option is generally useful when used inside emacs since it permits to
display the REPL prompts.
.IP "-n, --no-line-editor"
tell the REPL to not use the line editor even if GNU readline or BSD
libedit libraries are found. This permit to use an external tool such
as rlwrap(1) instead of the internal line editor support.
.IP "-d, --debug"
insert information to ease debugging (e.g, line numbers, signaling
usage of undefined symbols, ...). Can be used multiple times to
increase the debugging level;
.IP "-s, --stack-size=n"
use a stack of \fBn\fR cells.
.IP "--case-insensitive"
be case insensitive as in R5RS.
.IP "-c, --case-sensitive"
be case sensitive as in R7RS. This is the default value. This option
(and the previous one) permits to change the default behavior.
.IP "-u yes|no, --utf8-encoding=yes|no"
At initialisation time, \fBSTklos\fR infers the current locale from the
\fILC_ALL\fR variable. From the character encoding used in the selected
locale, \fBSTklos\fR determines if UTF-8 should be used for representing
strings.

The \fI--utf8-encoding\fR permits to force the encoding used. Note
that if \fILC_ALL\fR is unset and this option is not used, UTF-8
encoding will be used.

.IP "-F, --compiler-flags=list"
permits to set compiler flags to the to the values given in \fIlist\fR.
Boolean flags must begin by a '+' or a '-' character to indicate if the flag is set
or unset. Valued flags, must be followed by a '=' and the value given to the flag.
Flags are separated by the ',' character. For instance, the following
list \fI-F=+line-info,-time-display,unroll-iterations=5\fR sets the
\fIline-info\fR option, unsets the \fItime-display\fR option and initializes the
\fIunroll-iteration\fR to 5. The following flags can be used with
this option:
.IP "" 10
- \fIline-info\fR insert line numbers in the generated file. Setting this flag
is equivalent to the \fI-l\fR option. This flag uses the parameter
\fIcompiler:gen-line-number\fR.
.IP "" 10
- \fIshow-instructions\fR can be used to insert the textual instructions
into compiled files. This flag uses the parameter
\fIcompiler:show-assembly-code\fR.
.IP "" 10
- \fItime-display\fR can be used to display or not the time used to compile
a file. This flag uses the parameter
\fIcompiler:time-display\fR.
.IP "" 10
- \fIkeep-formals\fR can be used to control if formal arguments are stored in
user procedures. This flag uses the parameter
\fIcompiler:keep-formal\fR.
.IP "" 10
- \fIkeep-source\fR can be used to control if the source code is  stored in
user procedures. This flag uses the parameter
\fIcompiler:keep-source\fR.
.IP "" 10
- \fIinline-usuals\fR can be used to control if the usual functions are inlined
in produced code. This flag uses the parameter
\fIcompiler:inline-common-functions\fR.
.IP "" 10
- \fIsource-rewrite\fR can be used to control if the compiler tries to
rewrite/simplify source code before compiling it. This flag uses
the parameter \fIcompiler:source-rewrite\fR.
.IP "" 10
- \fIunroll-iterations\fR can be used to set the number of iterations to be unrolled
in loops. This flag uses the parameter \fIcompiler:unroll-iterations\fR and its
default value is 4.
.IP "" 10
- \fIpeephole-optimizer\fR can be used to activate or deactivate the peephole
optimizer during compilation. The peephole optimizer is used to change a
succession of instructions by another set of instructions that has better
performance. This flag uses the parameter \fIcompiler:peehole-optimizer\fR.
By default, the peephole optimizer is activated.
.IP "" 10
-\fIverify-assume\fR can be use to indicate if the compiler must include the
code necessary to verify the SRFI-145 assume forms, or not. This flag is false
by default, but it is forced to true if the debug mode is set.
.IP "-v, --version"
print program version and exit.
.IP "-V"
print detailed program version and information in machine-readable format,
as per SRFI-176, and exit.
.IP "-h, --help"
print this help and exit.
.IP "--"
This special option signals the end of the \fBstklos\fR command. The
arguments placed after this option are taken as program arguments,
even if they start with a "-" character.

.SH NOTES

Default case sensitivity depends on the way STklos is built.

.SH ENVIRONMENT VARIABLES

.IP "STKLOS_LOAD_PATH"
This is a colon-separated list of directories in which \fBstklos\fR
looks for loading files. It is used by functions such as \fBload\fR
or \fBtry-load\fR.

.IP "STKLOS_FRAMES"
This variable must contains an integer which indicates the number of
frames printed on an error. Use the value \fB0\fR for an unlimited
back-trace.

.IP "STKLOS_CONFDIR"
This variable can be used to designate the directory used by \fBstklos\fR
to store its configuration or packages files. If not set, the default
\fBstklos\fR configuration directory is by default \fB${XDG_CONFIG_HOME}/stklos\fR (or \fB~/.config/stklos/\fR if  \fBXDG_CONFIG_HOME\fR is unset).

.SH FILES

.IP "<stklos-conf-dir>/stklosrc"
This file is the user init file which is loaded when \fBstklos\fR
starts. Use the \fB-q\fR to avoid the loading of this file.


.SH "SEE ALSO"
.BR stklos-config (1),
.BR stklos-compile (1),
.BR stklos-pkg (1)

.LP
The full documentation for
.B stklos
is available at http://www.stklos.net/documentation.html


src/stklos.c
/*
 * stklos.c     -- STklos interpreter main function
 *
 * Copyright © 1999-2024 Erick Gallesio <eg@stklos.net>
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 *
 *           Author: Erick Gallesio [eg@unice.fr]
 *    Creation date: 28-Dec-1999 21:19 (eg)
 */

#include "stklos.h"
#include <langinfo.h>
#include "gnu-getopt.h"

#define ADD_OPTION(o, k)                                     do{\
  if (*o) options = STk_key_set(options,                        \
                                STk_makekey(k),                 \
                                STk_Cstring2string(o));         \
}while(0)

#define ADD_BOOL_OPTION(o, k)                        do{\
  options = STk_key_set(options,                        \
                        STk_makekey(k),                 \
                        MAKE_BOOLEAN(o));               \
}while(0)

#define ADD_INT_OPTION(o, k)                         do{\
  options = STk_key_set(options,                        \
                        STk_makekey(k),                 \
                        MAKE_INT(o));                   \
}while(0)

#define ADD_SCM_OPTION(o, k)                         do{\
  options = STk_key_set(options, STk_makekey(k),o);     \
}while(0)

/*=============================================================================
 *
 * Program arguments
 *
 *=============================================================================
 */

static char *boot_file    = "";
static char *program_file = "";
static char *load_file    = "";
static char *conf_dir     = "";
static char *sexpr        = "";
static char *cflags       = "";
static int  vanilla       = 0;
static int  fold_case     = 0;
static int  startup_msg   = 1;
static int  stack_size    = DEFAULT_STACK_SIZE;
static int  debug_mode    = 0;
static int  line_editor   = 1;
static int  srfi_176      = 0;
static char *script_file  = "";
static SCM  Idirs         = STk_nil;
static SCM  Adirs         = STk_nil;
static int  use_load_path = 0;


static struct option long_options [] =
{
  {"version",                no_argument,       NULL, 'v'},
  {"file",                   required_argument, NULL, 'f'},
  {"prepend-load-path",      required_argument, NULL, 'I'},
  {"append-load-path",       required_argument, NULL, 'A'},
  {"includes-use-load-path", no_argument,  NULL, 'y'},
  {"compiler-flags",         required_argument, NULL, 'F'},
  {"load",                   required_argument, NULL, 'l'},
  {"execute",                required_argument, NULL, 'e'},
  {"boot-file",              required_argument, NULL, 'b'},
  {"conf-dir",               required_argument, NULL, 'D'},
  {"no-init-file",           no_argument,       NULL, 'q'},
  {"no-startup-message",     no_argument,       NULL, 'Q'},
  {"interactive",            no_argument,       NULL, 'i'},
  {"no-line-editor",         no_argument,       NULL, 'n'},
  {"debug",                  no_argument,       NULL, 'd'},
  {"stack-size",             required_argument, NULL, 's'},
  {"case-sensitive",         no_argument,       NULL, 'c'},
  {"case-insensitive",       no_argument,       NULL, 'z'},
  {"utf8-encoding",          required_argument, NULL, 'u'},
  {"help",                   no_argument,       NULL, 'h'},
  {NULL,                     0,                 NULL, 0  }     /* for Clang */
};

static void SimpleVersion(void)
{
  printf("stklos %s\n", VERSION);
  printf("For more information, use the -V option.\n");
}

static void Usage(FILE *stream)
{
  fprintf(stream, "stklos %s\n", VERSION);
  fprintf(stream, "Usage: stklos [option ...] [--] [arg ... ]");
  fprintf(stream, "\n"
"Possible options:\n"
"   -l file, --load=file            load 'file' before going interactive\n"
"   -f file, --file=file            use 'file' as program\n"
"   -e sexpr, --execute=sexpr       evaluate the given sexpr and exit\n"
"   -b file, --boot-file=file       use 'file' to boot the system\n"
"   -D dir, --conf-dir=dir          change conf. dir (default: ~/.config/stklos)\n"
"   -I dir, --prepend-load-path=dir prepend 'dir' to the load path list\n"
"   -A dir, --append-load-path=dir  append 'dir' to the load path list\n"
"       --includes-use-load-path    include and include-ci use load-path\n"
"   -q, --no-init-file              quiet: do not load the user init file\n"
"   -Q, --no-startup-message        don't show the startup message\n"
"   -i, --interactive               interactive mode\n"
"   -n, --no-line-editor            don't use line editor\n"
"   -d, --debug                     add information to ease debugging\n"
"   -s, --stack-size=n              use a stack of size n (default %d)\n"
"   -c, --case-sensitive            be case sensitive by default\n"
"       --case-insensitive          be case insensitive by default\n"
"   -u, --utf8-encoding=yes|no      use/don't use UTF-8 encoding\n"
"   -F list, --compiler-flags=list  set/unset compiler flags (see below)\n"
"   -v, --version                   show version and exit (simple)\n"
"   -V                              show version and exit (detailed, SRFI-176)\n"
"   -h, --help                      show this help and exit\n"
"Compiler flags (option -F):\n"
"  Flags are separated by commas, booleans flags must be prefixed by '+' or  '-'\n"
"  Example:\n"
"      --compiler-flags='+line-info,-time-display,unroll-iterations=3'\n"
"  Possible flags:\n"
"    line-info          Insert line numbers in generated file (as -l option)\n"
"    show-instructions  Show instructions in generated file\n"
"    time-display       Print file compilation time\n"
"    keep-formals       Keep formal arguments in closures\n"
"    keep-source        Keep source code in closures\n"
"    inline-usuals      Inline usual functions\n"
"    allow-code-rewrite Permit code rewriting/simplification\n"
"    unroll-iterations  Set the number of iterations to be unrolled\n"
"    peephole-optimizer Use the peephole optimizer\n"
"All the arguments given after options are passed to the Scheme program.\n",
DEFAULT_STACK_SIZE);
}


static int process_program_arguments(int argc, char *argv[])
{
  extern char *optarg;
  extern int optind;
  int c;

  for ( ; ; ) {
    c = getopt_long(argc, argv, "+qQidnvVhcF:f:l:e:b:s:D:I:A:u:", long_options, NULL);
    if (c == -1) break;

    switch (c) {
      case 'v': SimpleVersion(); exit(0);
      case 'V': srfi_176        = 1;                                    break;
      case 'I': Idirs = STk_cons(STk_Cstring2string(optarg), Idirs);    break;
      case 'A': Adirs = STk_cons(STk_Cstring2string(optarg), Adirs);    break;
      case 'F': cflags          = optarg;                               break;
      case 'f': program_file    = optarg;                               break;
      case 'l': load_file       = optarg;                               break;
      case 'e': sexpr           = optarg;                               break;
      case 'b': boot_file       = optarg;                               break;
      case 'D': conf_dir        = optarg;                               break;
      case 'i': STk_interactive = 1;                                    break;
      case 'n': line_editor     = 0;                                    break;
      case 'd': debug_mode++;                                           break;
      case 'q': vanilla         = 1;                                    break;
      case 'Q': startup_msg     = 0;                                    break;
      case 's': stack_size      = atoi(optarg);                         break;
      case 'c': /* do nothing since it is the default now */            break;
      case 'y': use_load_path   = 1;                                    break;
      case 'z': fold_case       = 1;                                    break;
      case 'u': STk_use_utf8    = strspn(optarg, "yY1");                break;
      case 'h': Usage(stdout); exit(0);
      case '?': /* message error is printed by getopt */
                fprintf(stderr, "Try `%s --help' for more information\n", *argv);
                exit(1);
      default:  Usage(stderr); exit(1);
    }
  }
  return optind;
}

static void  build_scheme_args(int argc, char *argv[], char *argv0)
{
  SCM options, l = STk_nil;
  int i;

  for (i = argc-1; i >= 0; i--) {
    l = STk_cons(STk_Cstring2string(argv[i]), l);
  }
  options = LIST2(STk_makekey("argv"), l);
  ADD_OPTION("STklos",             "name");
  ADD_OPTION(argv0,                "program-name");
  ADD_OPTION(program_file,         "file");
  ADD_OPTION(load_file,            "load");
  ADD_OPTION(sexpr,                "sexpr");
  ADD_OPTION(conf_dir,             "conf-dir");
  ADD_OPTION(cflags,               "comp-flags");
  ADD_BOOL_OPTION(srfi_176,        "srfi-176");
  ADD_BOOL_OPTION(vanilla,         "no-init-file");
  ADD_BOOL_OPTION(startup_msg,     "startup-message");
  ADD_BOOL_OPTION(STk_interactive, "interactive");
  ADD_BOOL_OPTION(line_editor,     "line-editor");
  ADD_BOOL_OPTION(fold_case,       "fold-case");
  ADD_INT_OPTION(debug_mode,       "debug");
  ADD_BOOL_OPTION(STk_use_utf8,    "use-utf8");
  ADD_OPTION(script_file,          "script-file");
  ADD_SCM_OPTION(Idirs,            "prepend-dirs");
  ADD_SCM_OPTION(Adirs,            "append-dirs");
  ADD_BOOL_OPTION(use_load_path,   "use-load-path");
  STk_define_variable(STk_intern("*%system-state-plist*"), options,
                      STk_STklos_module);
}

int main(int argc, char *argv[])
{
  int ret;
  char *argv0 = *argv;

  /* Initialize the Garbage Collector */
#if (defined(__CYGWIN32__) &&  defined(GC_DLL)) || defined(_AIX)
# error GC problem
#endif
  STk_gc_init();

  /* Process command arguments */
  ret = process_program_arguments(argc, argv);
  argc -= ret;
  argv += ret;

  if (!*program_file && argc) {
    // We have at least one argument and we don't have set the -f option
    program_file = *argv;
    argv+=1;
    argc-=1;
  }

  if (*program_file) {
    if (strcmp(program_file, "-") == 0)
      /* Use stdin as input => reset program_file to "" */
      program_file = "";
    else
      /* Set script_file to for SRFI-196 */
      script_file = STk_expand_file_name(program_file);
  }

  /* See if we use UTF8 encoding */
  if (!setlocale(LC_ALL, "")) {
    fprintf(stderr, "Can't set the specified locale! "
            "Check LANG, LC_CTYPE, LC_ALL.\n");
    return 1;
  }

  if (STk_use_utf8 == -1) {
    /* user didn't force the encoding. Determine it from environment */
    STk_use_utf8 = (strcasecmp(nl_langinfo(CODESET), "UTF-8") == 0);
  }

  /* Hack: to give the illusion that there is no VM under the scene */
  if (*program_file) argv0 = program_file;

  /* Initialize the library */
  if (!STk_init_library(&argc, &argv, stack_size)) {
    fprintf(stderr, "cannot initialize the STklos library\nABORT\n");
    exit(1);
  }

  /* Place CLI arguments in the Scheme variable *%system-state-plist* */
  build_scheme_args(argc, argv, argv0);

  /* Boot the VM */
  if (*boot_file)
    /* We have a boot specified on the command line */
    ret = STk_load_boot(boot_file);
  else
    /* Use The boot in C file */
    ret = STk_boot_from_C();

  if (ret < 0) {
    fprintf(stderr, "cannot boot with \"%s\" file (code=%d)\n", boot_file, ret);
    exit(1);
  }

  STk_pre_exit(MAKE_INT(ret));
  return ret;
}
